### ISSUE TITLE
Compile error with rosencrantz 0.4 and --threads:on

### ISSUE THREAD

**Issue by aboisvert** [andreaferretti/rosencrantz#24]
Hi, I'm updating a project to Nim 1.0.0 and Rosencrantz 0.4 and I'm getting the following compile error,

```nim
       ... /home/boisvert/.nimble/pkgs/rosencrantz-0.4.0/rosencrantz/core.nim(75, 45) template/generic instantiation of `async` from here
        ... /home/boisvert/.choosenim/toolchains/nim-1.0.0/lib/pure/asyncmacro.nim(271, 31) Error: 'serverIter' is not GC-safe as it performs an indirect call via 'h'
```
Full compile trace:

```sh
~/n/t/server (master|✚2) $ nimble build --verbose
  Verifying dependencies for thermopi@0.1.0
    Reading official package list
   Checking for rosencrantz@0.4
      Info: Dependency on rosencrantz@0.4 already satisfied
  Verifying dependencies for rosencrantz@0.4.0
   Building thermopi/thermopi using c backend
     Error: Build failed for package: thermopi
        ... Details:
        ... Execution failed with exit code 1
        ... Command: "/home/boisvert/.nimble/bin/nim" c --noNimblePath -d:NimblePkgVersion=0.1.0 --path:"/home/boisvert/.nimble/pkgs/rosencrantz-0.4.0"  -o:"/home/boisvert/nim/thermopi/server/thermopi" "/home/boisvert/nim/thermopi/server/src/thermopi.nim"
        ... Output: Hint: used config file '/home/boisvert/.choosenim/toolchains/nim-1.0.0/config/nim.cfg' [Conf]
        ... Hint: used config file '/home/boisvert/nim/thermopi/server/nim.cfg' [Conf]
        ... Hint: system [Processing]
        ... Hint: widestrs [Processing]
        ... Hint: io [Processing]
        ... Hint: thermopi [Processing]
        ... Hint: asynchttpserver [Processing]
        ... Hint: tables [Processing]
        ... Hint: hashes [Processing]
        ... Hint: math [Processing]
        ... Hint: bitops [Processing]
        ... Hint: macros [Processing]
        ... Hint: algorithm [Processing]
        ... Hint: asyncnet [Processing]
        ... Hint: asyncdispatch [Processing]
        ... Hint: os [Processing]
        ... Hint: strutils [Processing]
        ... Hint: parseutils [Processing]
        ... Hint: unicode [Processing]
        ... Hint: pathnorm [Processing]
        ... Hint: osseps [Processing]
        ... Hint: posix [Processing]
        ... Hint: times [Processing]
        ... Hint: options [Processing]
        ... Hint: typetraits [Processing]
        ... Hint: heapqueue [Processing]
        ... Hint: lists [Processing]
        ... Hint: asyncstreams [Processing]
        ... Hint: asyncfutures [Processing]
        ... Hint: deques [Processing]
        ... Hint: cstrutils [Processing]
        ... Hint: monotimes [Processing]
        ... Hint: nativesockets [Processing]
        ... Hint: net [Processing]
        ... Hint: sets [Processing]
        ... Hint: selectors [Processing]
        ... Hint: epoll [Processing]
        ... Hint: uri [Processing]
        ... Hint: httpcore [Processing]
        ... Hint: threadpool [Processing]
        ... Hint: cpuinfo [Processing]
        ... Hint: linux [Processing]
        ... Hint: cpuload [Processing]
        ... Hint: rosencrantz [Processing]
        ... Hint: core [Processing]
        ... /home/boisvert/.nimble/pkgs/rosencrantz-0.4.0/rosencrantz/core.nim(75, 45) template/generic instantiation of `async` from here
        ... /home/boisvert/.choosenim/toolchains/nim-1.0.0/lib/pure/asyncmacro.nim(271, 31) Error: 'serverIter' is not GC-safe as it performs an indirect call via 'h'
```

It /looks/ like an issue with rosencrantz itself (since the compile error doesn't reference my code) but it's entirely possible that the issue is with my code.   

The calling code is open source is available here: https://github.com/aboisvert/thermopi/tree/master/server in case you want to look.  The rosencrantz specific code is here: https://github.com/aboisvert/thermopi/blob/master/server/src/thermopipkg/routes.nim

**Comment by aboisvert** [comment]
Interestingly, I see the same message when I build rosencrantz from source, however it is flagged as a `Warning`, not an `Error`.  

I wonder if this is because my application is compiled with `--threads:on`.

Details:

```sh
~/g/rosencrantz (master|✚1) $ nimble server
  Executing task server in /home/boisvert/git/rosencrantz/rosencrantz.nimble
  Verifying dependencies for rosencrantz@0.4.0
  Compiling tests/server (from package rosencrantz) using c backend
/home/boisvert/git/rosencrantz/rosencrantz/core.nim(75, 45) template/generic instantiation of `async` from here
/home/boisvert/.choosenim/toolchains/nim-1.0.0/lib/pure/asyncmacro.nim(271, 31) Warning: 'serverIter' is not GC-safe as it performs an indirect call via 'h' [GcUnsafe2]

proc innerProc(): auto =
  proc h(req: ref Request; ctx: Context): Future[Context] {.async.} =
    let x = req.url.path
    await req[].respond(Http200, x, {"Content-Type": "text/plain;charset=utf-8"}.newHttpHeaders)
    return ctx

  return h

innerProc()
```

**Comment by aboisvert** [comment]
As a temporary workaround, looks like I'm able to get over the compilation failure by using `--threadAnalysis:off`

**Comment by andreaferretti** [comment]
You are right that the issue is about Nim analysis of thread safety. Rosencrantz runs single threaded, but it is apparenly difficult to convice Nim of this. `--threadAnalysis:off` is a good workaround, but I will try to fix the underlying issue

**Comment by andreaferretti** [comment]
I published version 0.4.1, where I explicitly assert GC safety when building the main handler. Can you try that and see if it works for you?

**Comment by aboisvert** [comment]
Yes, that did it.  Compiling with v0.4.1 (and --threads:on) no longer
results in an error.  Thanks!!
